File: samples/java/Bingo/Server/Matchmaking/src/main/java/systems/zlink/samples/bingo/server/matchmaking/BingoMatchmaker.java
package systems.zlink.samples.bingo.server.matchmaking;

import systems.zlink.framework.spots.ZLinkInstanceSpot;
import systems.zlink.framework.spots.ZLinkInstanceSpotContext;

import java.time.Duration;
import java.time.Instant;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.atomic.AtomicInteger;

public final class BingoMatchmaker implements ZLinkInstanceSpot {
    private static final Duration IDLE_CHECK_PERIOD = Duration.ofSeconds(5);
    private static final Duration IDLE_CLOSE_AFTER = Duration.ofSeconds(30);
    private final ZLinkInstanceSpotContext context;
    private final AtomicInteger inFlight = new AtomicInteger();
    private volatile Instant lastActivity = Instant.now();

    public BingoMatchmaker(ZLinkInstanceSpotContext context) {
        this.context = context;
    }

    @Override
    public ZLinkInstanceSpotContext context() {
        return context;
    }

    @Override
    public CompletionStage<Void> onInitialize() {
        return context.addTimer(
                        "matchmaker-idle-close",
                        IDLE_CHECK_PERIOD,
                        BingoMatchmakerIdleTimerHandler.class,
                        null)
                .thenApply(ignored -> null);
    }

    void beginRequest() {
        inFlight.incrementAndGet();
        lastActivity = Instant.now();
    }

    void endRequest() {
        lastActivity = Instant.now();
        inFlight.decrementAndGet();
    }

    // --8<-- [start:doc-bingo-matchmaker-idle]
    void closeIfIdle() {
        if (inFlight.get() == 0
                && Duration.between(lastActivity, Instant.now()).compareTo(IDLE_CLOSE_AFTER) >= 0) {
            context.close();
        }
    }
    // --8<-- [end:doc-bingo-matchmaker-idle]

}

File: samples/kotlin/Bingo/Server/Matchmaking/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/matchmaking/BingoMatchmaker.kt
package systems.zlink.samples.kotlin.bingo.server.matchmaking

import java.time.Duration
import java.time.Instant
import java.util.concurrent.CompletionStage
import java.util.concurrent.atomic.AtomicInteger
import systems.zlink.framework.spots.ZLinkInstanceSpot
import systems.zlink.framework.spots.ZLinkInstanceSpotContext

class BingoMatchmaker(private val instanceContext: ZLinkInstanceSpotContext) : ZLinkInstanceSpot {
    private val inFlight = AtomicInteger()
    @Volatile private var lastActivity: Instant = Instant.now()

    override fun context(): ZLinkInstanceSpotContext = instanceContext

    override fun onInitialize(): CompletionStage<Void> =
        instanceContext
            .addTimer(
                "matchmaker-idle-close",
                IdleCheckPeriod,
                BingoMatchmakerIdleTimerHandler::class.java,
                null,
            )
            .thenApply { null }

    fun beginRequest() {
        inFlight.incrementAndGet()
        lastActivity = Instant.now()
    }

    fun endRequest() {
        lastActivity = Instant.now()
        inFlight.decrementAndGet()
    }

    // --8<-- [start:doc-bingo-matchmaker-idle]
    fun closeIfIdle() {
        if (
            inFlight.get() == 0 && Duration.between(lastActivity, Instant.now()) >= IdleCloseAfter
        ) {
            instanceContext.close()
        }
    }

    // --8<-- [end:doc-bingo-matchmaker-idle]

    private companion object {
        val IdleCheckPeriod: Duration = Duration.ofSeconds(5)
        val IdleCloseAfter: Duration = Duration.ofSeconds(30)
    }
}
